feat(chain): add `get` and `range` methods to `CheckPoint`

evanlinjin opened the pull request:
Partially fixes #1354 

### Description

These methods allow us to query for checkpoints contained within the linked list by height and height range. This is useful to determine checkpoints to fetch for chain sources without having to refer back to the `LocalChain`.

Currently this is not implemented efficiently, but in the future, we will change `CheckPoint` to use a skip list structure.

### Notes to the reviewers

Please refer to the conversation in #1354 for more details. In summary, both `TxGraph::missing_heights` and `tx_graph::ChangeSet::missing_heights_from` are problematic in their own ways. Additionally, it's a better API for chain sources if we only need to lock our receiving structures twice (once to obtain initial state and once for applying the update). Instead of relying on methods such as `EsploraExt::update_local_chain` to get relevant checkpoints, we can use these query methods instead. This allows up to get rid of `::missing_heights`-esc methods and remove the need to lock receiving structures in the middle of the sync.

### Changelog notice

* Added `get` and `range` methods to `CheckPoint` (and in turn, `LocalChain`). This simulates an API where we have implemented a skip list of checkpoints (to implement in the future). This is a better API because we can query for any height or height range with just a checkpoint tip instead of relying on a separate checkpoint index (which needs to live in `LocalChain`).
* Changed `LocalChain` to have a faster `Eq` implementation. We now maintain an xor value of all checkpoint block hashes. We compare this xor value to determine whether two chains are equal.
* Added `PartialEq` implementation for `CheckPoint` and `local_chain::Update`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

evanlinjin committed:
feat(chain): add `get` and `range` methods to `CheckPoint`

These methods allow us to query for checkpoints contained within the
linked list by height and height range. This is useful to determine
checkpoints to fetch for chain sources without having to refer back to
the `LocalChain`.

Currently this is not implemented efficiently, but in the future, we
will change `CheckPoint` to use a skip list structure.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -148,6 +149,34 @@ impl CheckPoint {
+    /// [`query_multi`]: CheckPoint::query_multi
+    pub fn query(&self, height: u32) -> Option<Self> {
+        self.iter()
+            .take_while(|cp| cp.height() >= height)
```
No need for this `take_while`.

evanlinjin replied:
I think we need this. If we don't have this and we don't have a direct hit, we will iterate the entire chain.

LLFourn replied:
Ah good point thanks! (comment that)

evanlinjin replied:
Comment added.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -148,6 +149,34 @@ impl CheckPoint {
+    ///
+    /// This returns an iterator which iterates over the requested heights and whether a
+    /// corresponding checkpoint is found.
+    pub fn query_multi(&self, heights: BTreeSet<u32>) -> impl Iterator<Item = (u32, Option<Self>)> {
```
Seems like a premature optimization. I understand that this is faster than calling query several times but the way to speed `query` up is with a skip list.

evanlinjin replied:
Even with a skip list, this will be faster? I also see this as a convenience method.

LLFourn replied:
> Even with a skip list, this will be faster? This is also a convenience method, like super convenient. I honestly think this is worth it.

Things that are faster are not necessarily better (if done before we have evidence that we need faster). Can we add the convenience method at the point that we need the convenience?

I think in general this PR doesn't really belong at this point because we don't get to see how we actually use the API. It seems like a simple thing to do as part of a larger PR that deletes the BTreeMap from LocalChain and uses checkpoint queries instead. Can we just make that this PR?

evanlinjin replied:
I realized that you are right. I think it is a great policy to only introduce convenience methods if they are used. `CheckPoint::query_multi` turns out to be not very useful. However, I introduced `CheckPoint::query_from` which will return `Some` if there is a checkpoint at `height` or greater. This is a sufficient replacement for `BTreeMap::<u32, BlockHash>::range` which is used in `bdk::Wallet::insert_tx`.

LLFourn reviewed (commented):
I'm slightly concerned we're introducing these methods without using them anywhere. Especially `query_multi` which I don't think needs to exist yet. I suggest cutting `query_multi`.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -205,18 +224,39 @@ pub struct Update {
 pub struct LocalChain {
     tip: CheckPoint,
-    index: BTreeMap<u32, BlockHash>,
+    xor: BlockHashXor,
```
I think this is a bit premature. I don't know that we need really fast equality for anything.  We didn't have fast equality before. If we were to do it, I'd just do it in `CheckPoint`.

evanlinjin replied:
> If we were to do it, I'd just do it in `CheckPoint`.

Do you mean implement fast equality in `CheckPoint`? So every checkpoint will have an `xor` field?

evanlinjin replied:
Yes, I agree that adding the xor right now is premature.

LLFourn replied:
> > If we were to do it, I'd just do it in `CheckPoint`.
> 
> Do you mean implement fast equality in `CheckPoint`? So every checkpoint will have an `xor` field?

Yes. As discussed in #1380 we probably do want to do this for `CheckPoint` in the future.

evanlinjin committed:
feat(chain): impl `PartialEq` on `CheckPoint`

We impl `PartialEq` on `CheckPoint` instead of directly on `LocalChain`.
We also made the implementation more efficient.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -34,6 +34,20 @@ struct CPInner {
 }
 
+impl PartialEq for CheckPoint {
+    fn eq(&self, other: &Self) -> bool {
```
:wrench: https://doc.rust-lang.org/core/iter/trait.Iterator.html#method.eq

I think it can be implemented with this.

ValuedMammal replied:
Would that not cause a recursion though?

LLFourn replied:
Oh interesting point. We still want to map to the `BlockId` so the eq used in the iterator is just `BlockId`'s. That should be fine!

evanlinjin replied:
Good point @ValuedMammal. I'll map the iterator to block id first.

evanlinjin replied:
Updated to use `Iterator::eq`.

LLFourn reviewed (approved):
ACK aa84ecf9d0a6f37e56c6fbebe3a2370f22f8acfd

ValuedMammal commented on crates/chain/tests/test_local_chain.rs:
```diff
@@ -528,6 +528,45 @@ fn checkpoint_from_block_ids() {
+            let exp_hash = t.chain.query(h).map(|cp| cp.hash());
+            match query_result {
+                Some(cp) => {
+                    assert_eq!(Some(cp.hash()), exp_hash);
```
It kind of looks like we're getting the checkpoint hash twice here from the same query result.

notmandatory replied:
I think this is OK, I read this line as saying: if the `query_result` is some checkpoint, then the hash of that checkpoint must equal the expected hash. 

⛏️ On the other hand I don't see how this test could ever fail since both `tip.query(h)` and `t.chain.query(h)` call `Checkpoint::query()` with `h`.

evanlinjin replied:
I realized how I wrote this was wrong. We shouldn't be using `query` for both what is to be checked and what we are checking against.

evanlinjin replied:
> I realized how I wrote this was wrong. We shouldn't be using `query` for both what is to be checked and what we are checking against.

Fixed.

notmandatory commented on crates/chain/tests/test_local_chain.rs:
```diff
@@ -528,6 +528,45 @@ fn checkpoint_from_block_ids() {
+                    assert_eq!(Some(cp.hash()), exp_hash);
+                    assert_eq!(cp.height(), h);
+                }
+                None => assert!(query_result.is_none()),
```
🔧 Since this match arm already told us `query_result` is `None` did you mean to do this instead?
```suggestion
                None => assert!(exp_hash.is_none()),
```

evanlinjin replied:
Fixed.

notmandatory reviewed (approved):
ACK aa84ecf9d0a6f37e56c6fbebe3a2370f22f8acfd

Appreciate the focused scope for this series of chain PRs, much easier to review. I have one small test code suggestion but otherwise looks good to me.

notmandatory reviewed (approved):
re-ACK 2b006da9042156b6ad9ce5a4237fc24973d0d72a

evanlinjin commented:
I would hold off merging this until this discussion reaches a consensus: https://github.com/bitcoindevkit/bdk/pull/1380#discussion_r1540554140

evanlinjin commented:
I have agreed with @LLFourn's suggestion of implementing `CheckPoint::range` and have updated the PR to implement it. Refer to the conversation here: https://github.com/bitcoindevkit/bdk/pull/1380#discussion_r1547182439

This PR is ready for review/merge.

evanlinjin commented:
Ahhh I probably still need to add a few tests.

jirijakes commented on crates/chain/src/local_chain.rs:
```diff
@@ -148,6 +158,39 @@ impl CheckPoint {
+        self.iter()
+            // optimization to avoid iterating the entire chain if we do not get a direct hit
+            .take_while(|cp| cp.height() >= height)
+            .find(|cp| cp.height() == height)
```
Perhaps opportunity to reuse `range`, since `query` is its special case? It seems that even performance-wise it would not be worse.

```suggestion
	self.range(height..=height).next()
```

evanlinjin replied:
Changed.

jirijakes commented on crates/chain/src/local_chain.rs:
```diff
@@ -148,6 +158,39 @@ impl CheckPoint {
+    /// Query for checkpoint at `height`.
+    ///
+    /// Returns `None` if checkpoint at `height` does not exist`.
+    pub fn query(&self, height: u32) -> Option<Self> {
```
The name feels hard to discover. If I am interested in finding a checkpoint at a _height_, I would first look for a method containing _height_, something like `find_at_height()`, `find_by_height()`, `get_at_height()`, `at_height()` or so.

Similarly `range` and the equivalent methods on `LocalChain`.

evanlinjin replied:
What about `query_height`? cc @LLFourn

LLFourn replied:
I'd go with just `.get_hash` or better yet just `get`. `query` is not a good name indeed.

evanlinjin replied:
Let's go with `.get`.

LLFourn commented on crates/chain/src/local_chain.rs:
```diff
@@ -148,6 +158,39 @@ impl CheckPoint {
+
+    /// Iterate checkpoints over a height range.
+    ///
+    /// Note that we always iterate checkpoints in reverse height order (iteration starts at tip
```
:pick: there's no need to start this sentence with `Note that`. It doesn't start at the tip it starts at the highest height within the range.

The implementation looks correct fwiw. I still feel like we are adding methods like this before we actually need them. My bias is to keep it for now since you've done it but we really have to stop adding these kinds of methods.

evanlinjin replied:
This is needed though, like we're actually using it

LLFourn replied:
I see one place it's being used where [as we've discussed](https://github.com/bitcoindevkit/bdk/pull/1380#discussion_r1540554140) it's not necessary.

evanlinjin replied:
Conversation continues here: https://github.com/bitcoindevkit/bdk/pull/1369#discussion_r1550753230

ValuedMammal commented on crates/chain/src/local_chain.rs:
```diff
@@ -202,21 +245,19 @@ pub struct Update {
-    }
 }
 
+// TODO: Figure out whether we can get rid of this
```
You could make this be the implementation for the `LocalChain::blocks` function and have it return `BTreeMap<u32, BlockHash>`, possibly even taking an extra `range: impl RangeBounds<u32>` parameter.

evanlinjin replied:
This is an interesting idea! However, I like to use @LLFourn's policy when it comes to introducing helper methods/structures. We should only introduce such things if we introduce how/where it is used within the PR. This ensures the usefulness of the things we are introducing.

I've decided to remove this `impl` completely as it was not being used.

ValuedMammal commented on crates/chain/src/local_chain.rs:
```diff
@@ -1,6 +1,8 @@
 
 use core::convert::Infallible;
+use core::ops::RangeBounds;
+use core::u32;
```
Is this `core::u32` needed?

evanlinjin replied:
I blame my autocomplete engine.

evanlinjin replied:
Removed

evanlinjin replied:
Removed.

ValuedMammal reviewed (commented):
Overall looks good.

`get` sounds like a better name than `query`. My other suggestion would have been to call it `find`, since that is how it's implemented.

Interesting that `LocalChain` now delegates virtually everything to the checkpoint. Curious to see if the structure of `LocalChain` becomes more relevant when skip list is involved.

evanlinjin commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1128,18 +1128,14 @@ impl<D> Wallet<D> {
-                    .blocks()
                     .range(height..)
-                    .next()
+                    .last()
```
@LLFourn `CheckPoint::range` is also used here.

LLFourn replied:
Thanks. I missed that one and had forgot about the API of this function. On a side note this is interesting because we are anchoring a tx not to the block that it's in. So even if we made a rule that esplora/electrum etc must tell us about the block that the tx is in we'd still need the concept of anchoring if we want to keep this API.

LLFourn commented:
ACK cdbc131ef68182a4317beb145232ec8e87d6299c

evanlinjin committed:
chore(chain)!: rm `From<LocalChain> for BTreeMap<u32, BlockHash>`

I don't think this was ever used. The only possible usecase I can think
of is for tests, but I don't think that is a strong enough incentive for
us to keep this.

evanlinjin commented:
I've addressed all review comments and removed the `From<LocalChain>` impl for `BTreeMap<u32, BlockHash>` as it wasn't being used.

jirijakes commented on crates/chain/src/local_chain.rs:
```diff
@@ -276,39 +299,30 @@ impl LocalChain {
-
-        if chain.index.get(&0).copied().is_none() {
+        let genesis_cp = tip.iter().last().expect("must have at least one element");
+        if genesis_cp.height() != 0 {
             return Err(MissingGenesisError);
         }
-
-        debug_assert!(chain._check_index_is_consistent_with_tip());
-        Ok(chain)
+        Ok(Self { tip })
     }
 
     /// Constructs a [`LocalChain`] from a [`BTreeMap`] of height to [`BlockHash`].
     ///
     /// The [`BTreeMap`] enforces the height order. However, the caller must ensure the blocks are
     /// all of the same chain.
     pub fn from_blocks(blocks: BTreeMap<u32, BlockHash>) -> Result<Self, MissingGenesisError> {
-        if !blocks.contains_key(&0) {
+        if blocks.get(&0).is_none() {
```
Just curious: why this change? Original looked a bit better readable.

evanlinjin replied:
`.contains_key` no longer exists because `blocks` is a different type now.

jirijakes commented on crates/chain/src/local_chain.rs:
```diff
@@ -542,33 +550,41 @@ impl LocalChain {
+                remove_from = Some(cp);
+            }
         }
+        self.tip = match remove_from.map(|cp| cp.prev()) {
+            // The checkpoint below the earliest checkpoint to remove will be the new tip.
+            Some(Some(new_tip)) => new_tip,
+            // If there is no checkpoint below the earliest checkpoint to remove, it means the
+            // "earliest checkpoint to remove" is the genesis block. We disallow removing the
+            // genesis block.
+            Some(None) => return Err(MissingGenesisError),
+            // If there is nothing to remove, we return an empty changeset.
+            None => return Ok(ChangeSet::default()),
+        };
+        Ok(changeset)
```
Not too important, but it took me a while to understand what changeset is returned when and what is when assigned to `self.tip` and when it is left untouched.

Perhaps?

```suggestion
        match remove_from.map(|cp| cp.prev()) {
            // The checkpoint below the earliest checkpoint to remove will be the new tip.
            Some(Some(new_tip)) => {
                self.tip = new_tip;
                Ok(changeset)
            }
            // If there is no checkpoint below the earliest checkpoint to remove, it means the
            // "earliest checkpoint to remove" is the genesis block. We disallow removing the
            // genesis block.
            Some(None) => Err(MissingGenesisError),
            // If there is nothing to remove, we return an empty changeset.
            None => Ok(ChangeSet::default()),
        }
```

evanlinjin replied:
Your suggestion is definitely better, but I'll ignore it now because I want this merged and rebasing is annoying 😅

jirijakes reviewed (approved):
The PR helped me better understand some internals. Still learning.

Only have one question and one minor suggestion (not too important).

ACK 6d78f0e0fbaf7b25499ed671fdd2c1266a0d47c7

LLFourn reviewed (approved):
ACK 53942cced492138174638a8087b08f643a8d41ad
